#7091 support socks5

huisongma opened the pull request:
support socks5 proxy

huisongma committed:
#7091 support socks5

huisongma committed:
add license head

huisongma committed:
Update Socks5Proxy.java

code format

huisongma committed:
Update Socks5Proxy.java

code format

huisongma committed:
Update Socks5Proxy.java

code format

sbordet commented:
Thanks for this contribution!

Since SOCKS5 is way more complicated than SOCKS4, I would like to avoid hardcoding the authentication mechanism.
The idea is to have a `Socks5` class, with inner classes that implement different authentication mechanisms, for now only "no authentication" (say `Socks5.NoAuthentication`) and "username/password" (say `Socks5.UsernamePasswordAuthentication`).

Shared code can be put in the `Socks5` class.

Also, test cases are missing.
I would like to see at least test cases for abrupt connection close from the server,  no answer from the server (timeout on client), garbage bytes from the server, response failure from server and response succeeded from server.

Please try to stick with Java/Jetty conventions. For example, `enum` constants are all upper case, the `log` instance is upper case, etc.

huisongma committed:
Add Socks5.java

1.add Authorization interface
2.add NoAuthorization and UsernamePasswodAuthorization implements

huisongma committed:
add test case

Commit:
change code style

huisongma commented:
continue on #9653
